rfcs/2022-02-24-11552-dd-agent-style-secret-management.md

```diff
@@ -0,0 +1,182 @@
+
+Datadog Secret API, as per the official [doc][dd-secret-backend-exec-api]:
+
+- Communication happens on stdout/sdin
+- It uses plain text json
+
+Vector would run the user-provided executable and feed its standard input with the list of secrets to retrieve:
+
+```json
+{"version": "1.0", "secrets": ["secret1", "secret2"]}
+```
+
+The version used by the Datadog agent is 1.0. The user provided executable should then reply on its standard output with
+a JSON formatted string:
+
+```json
+{
+  "secret1": {"value": "secret_value", "error": null},
+  "secret2": {"value": null, "error": "could not fetch the secret"}
+}
+```
+
+**Note**: The version field can be used to introduce specific behaviour, one major useful thing that could be introduced
+is the ability for the user provided executable to provide a secrete expiration date. This is mentioned in the
+improvements section.
+
+New top level options to be added will be sitting inside the `secret` namespace, this would lead to something like:
+
+```toml
+[secret.local_exec]
+type = "exec"
+path = "/path/to/the/command"
+argument = "--config foo=bar"
+timeout = 5
+
+[secret.prod_vault]
+type = "vault"
+address = "https://vault.corp.tld/"
+token = "secret://local_exec/vault_token"
+timeout = 5
+
+[sources.system_logs]
+type = "file"
+includes = ["/var/log/system.log"]
+
+[sinks.app_logs]
+type = "datadog_logs"
+default_api_key = "secret://prod_vault/dd_api_2022_02"
+inputs = ["system_logs"]
+```
+
+The first implementation would only support the `exec` backend but with extensibility point clearly identified to easily
+implement additional backend if needed.
+
+Overall the behaviour for corner cases should follow what's in place for environmment variable interpolation as this is
+a very close feature. When a configuration reload happens, secrets shall also be refreshed.
+
+### Implementation
+
+A secret backend that would lie in `./src/config/secret.rs` and would call the user provided executable for secrets,
+cache secrets to avoid calling the backend for the same key multiple time. It would read the configuration file once to
+get its config before further processing, ideally env var interpolating should be supported (this should not be a
+problem). In `./src/config/builder.rs`, `load_builder_from_paths` will still be returning a complete configuration with
+placeholders replaced by secrets.
+
+The `ConfigBuilder` struct will get a new `secret_backend` field (type to something like `SecretBackend`), this means
+that `load_builder_from_paths` will then assert if this field is present before returning to the caller, and if it, the
+config will be reloaded with this `SecretBackend` passed to downstream callee and hook the secret interpolation around
+the same point as [the environment variable interpolation][env-var-hook] it would then query this `SecretBackend` for
+each `ENC[secret_key]`, `ENC[backend_name/secret_key]` or `secret://backend_name/secret_key` (The placeholded choice
```

Comment:
I think I prefer one of the options that stands out more than `secret://backend_name/secret_key`. If we did `secret://backend_name/secret_key` I might still like to see it wrapped in something to call attention to it.

Maybe: `${$secret.backend_name.secret_key}`? This collides with environment variable interpolation, but that seems like a reasonable restriction to me. I'm not seeing what `ENC` is supposed to mean. I'm guessing "encrypted", but the secrets could be coming from anywhere and are not necessarily encrypted.

Just `SECRET[backend_name/secret_key]` seems fine too given the Datadog Agent uses it and so presumably they've thought about the different interpolation options already.

Reply:
I agree with @jszwedko . Also, why bother name spacing secrets with the backend it came from? At the end of the day it shouldn't matter where it came from, so I'd prefer just `SECRET[secret_key]`, similar to environment variables.

Reply:
I made the whole thing a bit more detailed, this should clarify how the implementation is foreseen.

Reply:
Thanks @prognant, I understand that multiple backends could have a secret name conflict, but I fundamentally disagree that referencing a secret should be coupled with the backend. IMO, the Vector configuration should only describe the secret it wants and not also where it came from. This way secrets can easily move between backends without changing the Vector configuration.

Reply:
I agree that this would make the configuration less verbose without comprise on its readability. However this would imply querying all backends for more or less all secrets that need to be retrieved, while this is not really a problem, this imply carefully documenting and implementing backend precedence (probably in the order they appear in config). Thinking a bit more about that perspective this raise the point of inverting the logic (thus not following the existing Datadog API); i.e. Vector would blindly query all backends, they would reply with all the secrets they know then Vector would check that it got all the secret needed and interpolate the configuration. That being said I still prefer the query logic where the information flow is limited to what's mandatory.

Reply:
I'm confused, why would someone configure a backend if they aren't going to use the secrets in the Vector configuration file?